File: src/octave/dd_linearsolv/inst/seidel.m
## Copyright (C) 2007 Dominik Dahlem <Dominik.Dahlem@cs.tcd.ie>
##  
## This file is free software; as a special exception the author gives
## unlimited permission to copy and/or distribute it, with or without 
## modifications, as long as this notice is preserved.
## 
## This program is distributed in the hope that it will be useful, but
## WITHOUT ANY WARRANTY, to the extent permitted by law; without even the
## implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

## -*- texinfo -*-
## @deftypefn{Function File} {[@var{x_bar}, @var{x_error}, @var{max_error}] = }
##     dd_seidel (@var{A}, @var{b}, @var{x})
## @cindex gauss-seidel method, Iterative Method
##
## @var{A} is the matrix to be solved with the solution vector @var{b}
## and a given x vector @var{x}.
## @var{x_bar} is the solution.
##
## @seealso{dd_gauss}
## @end deftypefn


function [x_bar, x_error, max_error] = dd_seidel(A, b, x)
  if (nargin != 3)
    usage("dd_chol(A, b, x)");
  endif

  if (matrix_type(A) != "Positive Definite")
    error("The matrix A has to be positive definite.");
  endif

  if (rows(b) != rows(A))
    error("The vector b has to have the same dimension as rows in matrix A");
  endif

  if (!isvector(b))
    error("b must be an array.");
  endif

  x_bar = x;
  x_old = x;
  
  for l = 1:rows(A)
    for i = 1:rows(A)
      temp1 = 0;
      temp2 = 0;
      
      for j = 1:columns(A)
	if (j < i)
	  temp1 += A(i, j) * x_bar(j);
	elseif (j > i)
	  temp2 += A(i, j) * x_old(j);
	endif
      endfor
      
      x_bar(i) = (b(i) - temp1 - temp2) / A(i, i);
    endfor

    x_old = x_bar;
  endfor
  
  x_error = x - x_bar;
  max_error = max(abs(x_error));
  
endfunction

#!test
#! A = [1.86279, 0.47863, -0.54877; 0.47863, 1.61609, 0.10628; -0.54877,
#!      0.10628, 2.76115]
#! b = [1;2;3]
#! x = [0;0;0]
#! [x_bar, x_error, max_error] = dd_seidel(A, b, x)
#! assert(A * x_bar, b, 0.001)

File: src/c/solvers/seidel.c
/* Copyright (C) 2007 Dominik Dahlem <Dominik.Dahlem@gmail.com>                */
/*                                                                             */
/* This file is free software; as a special exception the author gives         */
/* unlimited permission to copy and/or distribute it, with or without          */
/* modifications, as long as this notice is preserved.                         */
/*                                                                             */
/* This program is distributed in the hope that it will be useful, but         */
/* WITHOUT ANY WARRANTY, to the extent permitted by law; without even the      */
/* implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.    */
#include <gsl/gsl_matrix.h>
#include <gsl/gsl_vector.h>

#include "seidel.h"
#include "error.h"
#include "util.h"



int seidel(gsl_matrix *A, gsl_vector *b, gsl_vector *x,
           gsl_vector **x_bar, gsl_vector **x_error, double *max_error)
{
    gsl_vector *x_old;
    size_t l, i, j;
    double temp1, temp2;
    

    if (b->size != A->size1) {
        return MATRIX_VECTOR_UNEQUAL_ROW_DIM;
    }
    
/*     if ((status = isPositiveDefinite(A))) { */
/*         return status; */
/*     } */

    *x_bar = gsl_vector_alloc(x->size);
    x_old = gsl_vector_alloc(x->size);
    gsl_vector_memcpy(*x_bar, x);
    gsl_vector_memcpy(x_old, x);
    
    for (l = 0; l < A->size1; ++l) {
        for (i = 0; i < A->size1; ++i) {
            temp1 = 0.0;
            temp2 = 0.0;
            
            for (j = 0; j < A->size2; ++j) {
                if (i > j) {
                    temp1 += gsl_matrix_get(A, i, j) * gsl_vector_get(*x_bar, j);
                } else if (i < j) {
                    temp2 += gsl_matrix_get(A, i, j) * gsl_vector_get(x_old, j);
                }
            }

            gsl_vector_set(*x_bar, i, (gsl_vector_get(b, i) - temp1 - temp2) /
                           gsl_matrix_get(A, i, i));
        }

        gsl_vector_memcpy(x_old, *x_bar);
    }

    /* get the error statistics */
    gatherErrors(*x_bar, x, x_error, max_error);

    gsl_vector_free(x_old);
    
    return 0;
}
